### ISSUE TITLE
Understanding the optimizers get_updates routine

### ISSUE THREAD

**Issue by mha-py** [keras-team/keras#11937]
Hi!

I would like to adapt the adam routine (to effectively compute larger batches than fit into the vram), but theres something in the get_updates routine that i really do not get....
The code in sgd for example is is:
```
        shapes = [K.int_shape(p) for p in params]
        moments = [K.zeros(shape) for shape in shapes]      ### moments are set zero
        self.weights = [self.iterations] + moments
        for p, g, m in zip(params, grads, moments):         ### m loops over all moments
            v = self.momentum * m - lr * g  # velocity      ### m is read out? why when its zero?
            self.updates.append(K.update(m, v))             ### m will be updated but never read out?
            if self.nesterov:
                new_p = p + self.momentum * v - lr * g
            else:
                new_p = p + v
            # Apply constraints.
            if getattr(p, 'constraint', None) is not None:
                new_p = p.constraint(new_p)
        self.updates.append(K.update(p, new_p))
```

It looks like the variables in "momentum" (in the for loop "m") are initiated as zero each time the function is called. Then the new momentum "v" is put into "m", but in the value is thrown away next time. I dont get it, really.... Could someone enlight me? I feel like a newbie...

**Comment by BadrYoubiIdrissi** [comment]
I think get_updates returns a list of backend operations (updates) to apply to the tensors for a single pass of gradient descent. However these updates are only calculated once and then applied multiple time! So you only call once (and only once) get_updates :

```
opt = keras.optimizers.Adam()
loss = K.mean(input)
updates = opt.get_updates(loss, params)
iterateFunction = K.function([input],[loss], updates=updates)
for i in range(100):
    iterateFunction([inputTensor])
```

**Comment by mha-py** [comment]
Yeah I know that, my problem was I would like to modify the adam optimizer such that it is storing the gradients for some update calls before actually updating. So i want to understand why the implementation of internal adam parameters (like the momentum of each parameter) is working like that
